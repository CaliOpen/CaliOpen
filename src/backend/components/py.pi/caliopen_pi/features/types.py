# -*- coding: utf-8 -*-
"""Caliopen privacy features definition."""

from __future__ import unicode_literals, absolute_import

import logging

# Privacy features definition
FEATURES_0 = {'technic': [{'mail_emitter_mx_reputation': {'type': 'string'}},
                          {'mail_emitter_certificate': {'type': 'string'}},
                          {'mail_agent': {'type': 'string'}},
                          {'transport_signed': {'type': 'bool'}},
                          {'message_signed': {'type': 'bool'}},
                          {'message_signature_type': {'type': 'string'}},
                          {'message_encrypted': {'type': 'bool'}},
                          {'message_encryption_infos': {'type': 'string'}},
                          {'is_spam': {'type': 'bool'}},
                          {'spam_score': {'max': 100,
                                          'min': 0,
                                          'type': 'int'}},
                          {'spam_method': {'type': 'string'}},
                          {'ingress_socket_version': {'type': 'string'}},
                          {'ingress_cipher': {'type': 'string'}},
                          {'nb_external_hops': {'type': 'int'}}]}

FEATURE_DIMENSIONS = FEATURES_0.keys()

log = logging.getLogger(__name__)


def check_feature_bounding(value, feature):
    """Check that a feature fit between min and max value if apply."""
    if 'min' in feature and value < feature['min']:
        return False
    if 'max' in feature and value > feature['max']:
        return False
    return True


def check_feature(name, value):
    """Check that a feature have a correct value."""
    for dimension in FEATURE_DIMENSIONS:
        if name in FEATURES_0[dimension]:
            feature = FEATURES_0[dimension][name]
            break
    else:
        return False
    if feature['type'] == 'int':
        try:
            int(value)
        except ValueError:
            return False
        return check_feature_bounding(value, feature)

    elif feature['type'] == 'bool':
        return True if value in (True, False) else False


def marshall_feature(name, value):
    """Marshall a feature to it's type representation."""
    if not check_feature(name, value):
        raise ValueError('Invalid value {} for feature {}'.format(value, name))
    feature = FEATURES_0[name]
    if feature['type'] == 'int':
        return int(value)
    elif feature['type'] == 'bool':
        return True if value.lower().startswith('t') else False


def marshall_features(features):
    """Marshall a dict of features for suitable output."""
    res = {}
    for name, value in features.items():
        try:
            new_value = marshall_feature(name, value)
            res[name] = new_value
        except ValueError:
            log.warn('Feature {} with {} do not marshall'.format(name, value))
            pass
    return res


def unmarshall_features(features):
    """Unmarshall a dict of features suitable for storage."""
    res = {}
    for k, v in features.items():
        if not (v == '' or v is None):
            res[k] = str(v)
    return res
