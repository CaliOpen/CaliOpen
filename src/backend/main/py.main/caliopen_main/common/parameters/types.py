# -*- coding: utf-8 -*-
"""Caliopen contact parameter validators."""
from __future__ import absolute_import, print_function, unicode_literals

from schematics.types import StringType
from schematics.exceptions import ValidationError
import phonenumbers

from ..helpers.normalize import clean_email_address


class InternetAddressType(StringType):

    """Validate an email or instant messaging address, return normalized."""

    def validate_email(self, value, context=None):
        try:
            clean, email = clean_email_address(value)
        except Exception as exc:
            raise ValidationError(exc.message)
        return clean


class PhoneNumberType(StringType):

    """Validate a phone number and normalize in international format."""

    def validate_phone(self, value, context=None):
        try:
            number = phonenumbers.parse(value, None)
            phone_format = phonenumbers.PhoneNumberFormat.INTERNATIONAL
            return phonenumbers.format_number(number, phone_format)
        except Exception as exc:
            raise ValidationError(exc.message)
