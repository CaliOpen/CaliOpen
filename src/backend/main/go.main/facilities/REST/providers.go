package REST

import (
	"context"
	"errors"
	"fmt"
	. "github.com/CaliOpen/Caliopen/src/backend/defs/go-objects"
	"github.com/CaliOpen/Caliopen/src/backend/main/go.main/users"
	"github.com/CaliOpen/go-twitter/twitter"
	"github.com/dghubble/oauth1"
	twitterOAuth1 "github.com/dghubble/oauth1/twitter"
	"github.com/satori/go.uuid"
	"golang.org/x/oauth2"
	googleApi "google.golang.org/api/oauth2/v2"
	"net/http"
	"time"
)

func (rest *RESTfacility) RetrieveProvidersList() (providers []Provider, err error) {
	if rest.providers != nil {
		providers = []Provider{}
		for _, provider := range rest.providers {
			providers = append(providers, provider)
		}
		return providers, nil
	}
	return providers, errors.New("providers slice is nil")
}

// GetProviderOauthFor returns provider's params required for authenticated user to initiate an Oauth request
// In case of Twitter, auth request url is fetched from twitter API endpoint on the fly.
// For all requests, an Oauth session cache is initialized for requesting user, making use of cache facility.
func (rest *RESTfacility) GetProviderOauthFor(userId, name string) (provider Provider, err CaliopenError) {
	provider, found := rest.providers[name]
	if found {
		switch provider.Name {
		case "twitter":
			requestToken, requestSecret, e := setTwitterAuthRequestUrl(&provider, rest.Hostname)
			if e != nil {
				err = WrapCaliopenErrf(e, FailDependencyCaliopenErr, "[GetProviderOauthFor] failed to set twitter aut request")
				return
			}
			rest.Cache.SetOauthSession(requestToken, &OauthSession{
				RequestSecret: requestSecret,
				UserId:        userId,
			})
		case "gmail":
			state, e := users.SetGoogleAuthRequestUrl(&provider, rest.Hostname)
			if e != nil {
				err = WrapCaliopenErrf(e, FailDependencyCaliopenErr, "[GetProviderOauthFor] failed to set twitter aut request")
				return
			}
			rest.Cache.SetOauthSession(state, &OauthSession{
				UserId: userId,
			})
		default:
			err = NewCaliopenErr(NotImplementedCaliopenErr, "not implemented")
			return
		}
		return
	}
	err = NewCaliopenErr(NotFoundCaliopenErr, "not found")
	return
}

func (rest *RESTfacility) CreateTwitterIdentity(requestToken, verifier string) (remoteId string, err CaliopenError) {
	provider := rest.providers[TwitterProtocol]
	conf := &oauth1.Config{
		ConsumerKey:    provider.Infos["consumer_key"],
		ConsumerSecret: provider.Infos["consumer_secret"],
		Endpoint:       twitterOAuth1.AuthorizeEndpoint,
	}
	oauthCache, e := rest.Cache.GetOauthSession(requestToken)
	if e != nil {
		err = WrapCaliopenErrf(e, NotFoundCaliopenErr, "[CreateTwitterIdentity] failed to retrieve Oauth session in cache for request token %s", requestToken)
		return
	}
	accessToken, accessSecret, e := conf.AccessToken(requestToken, oauthCache.RequestSecret, verifier)
	if e != nil {
		err = WrapCaliopenErrf(e, FailDependencyCaliopenErr, "[CreateTwitterIdentity] AccessToken request failed")
	}

	// retrieve twitter profile from Twitter api
	token := oauth1.NewToken(accessToken, accessSecret)
	httpClient := conf.Client(oauth1.NoContext, token)
	twitterClient := twitter.NewClient(httpClient)
	accountVerifyParams := &twitter.AccountVerifyParams{
		IncludeEntities: twitter.Bool(false),
		SkipStatus:      twitter.Bool(true),
		IncludeEmail:    twitter.Bool(false),
	}
	twitterUser, resp, e := twitterClient.Accounts.VerifyCredentials(accountVerifyParams)
	if e != nil || resp.StatusCode != http.StatusOK ||
		twitterUser == nil || twitterUser.ID == 0 || twitterUser.IDStr == "" {
		err = NewCaliopenErr(FailDependencyCaliopenErr, "[CreateTwitterIdentity] twitter client failed to get Twitter User")
		return
	}

	// build user identity
	//1.check if this user_identity already exists
	foundIdentities, e := rest.store.LookupIdentityByIdentifier(twitterUser.ScreenName, TwitterProtocol)
	if e != nil {
		err = WrapCaliopenErrf(e, DbCaliopenErr, "[CreateTwitterIdentity] failed to lookup in store if identity already exists. Aborting")
		return
	}
	foundCount := len(foundIdentities)
	switch foundCount {
	case 0:
		userIdentity := new(UserIdentity)
		userID := UUID(uuid.FromStringOrNil(oauthCache.UserId))
		userIdentity.MarshallNew(userID)
		userIdentity.Protocol = TwitterProtocol
		userIdentity.Type = RemoteIdentity
		userIdentity.DisplayName = twitterUser.Name
		userIdentity.Identifier = twitterUser.ScreenName
		userIdentity.Credentials = &Credentials{
			"token":  accessToken,
			"secret": accessSecret,
		}
		userIdentity.Infos = map[string]string{
			"provider": "twitter",
			"authtype": Oauth1,
		}
		// save identity
		e := rest.CreateUserIdentity(userIdentity)
		if e != nil {
			err = WrapCaliopenErr(e, FailDependencyCaliopenErr, "[CreateTwitterIdentity] failed to create user identity")
			return
		}
		remoteId = userIdentity.Id.String()
		return
	case 1:
		// this twitter identity already exists, checking if it belongs to this user and, if ok, just updating Twitter's name
		storedIdentity, e := rest.RetrieveUserIdentity(foundIdentities[0][0], foundIdentities[0][1], false)
		if e != nil || storedIdentity == nil {
			err = WrapCaliopenErrf(e, DbCaliopenErr, "[CreateTwitterIdentity] failed to retrieve user identity found for twitter account %s", twitterUser.ScreenName)
			return
		}
		modifiedFields := map[string]interface{}{
			"DisplayName": twitterUser.Name,
		}
		storedIdentity.DisplayName = twitterUser.Name
		if e := rest.store.UpdateUserIdentity(storedIdentity, modifiedFields); e != nil {
			err = WrapCaliopenErrf(e, FailDependencyCaliopenErr, "[CreateTwitterIdentity] failed to update user identity in db")
			return
		}
		remoteId = storedIdentity.Id.String()
		return
	default:
		err = NewCaliopenErrf(FailDependencyCaliopenErr, "[CreateTwitterIdentity] inconsistency in store : more than one identity found with twitter screen name <%s>", twitterUser.ScreenName)
		return
	}
}

func (rest *RESTfacility) CreateGmailIdentity(state, code string) (remoteId string, err CaliopenError) {

	// start by checking if we have the state in cache
	oauthCache, e := rest.Cache.GetOauthSession(state)
	if e != nil {
		err = WrapCaliopenErrf(e, NotFoundCaliopenErr, "[CreateGmailIdentity] failed to retrieve Oauth session in cache for state %s", state)
		return
	}

	gmailProvider := rest.providers["gmail"]
	gmailProvider.OauthCallbackUri = fmt.Sprintf(users.CALLBACK_BASE_URI, "gmail")

	oauthConfig := users.SetGoogleOauthConfig(gmailProvider, rest.Hostname)

	ctx := context.TODO()

	// get access & refresh tokens that will be added to UserIdentity.Credentials
	token, e := oauthConfig.Exchange(ctx, code, oauth2.AccessTypeOffline)
	if e != nil {
		err = WrapCaliopenErrf(e, NotFoundCaliopenErr, "[CreateGmailIdentity] failed to retrieve access token for state %s", state)
		return
	}
	if token.RefreshToken == "" {
		err = WrapCaliopenErrf(e, NotFoundCaliopenErr, "[CreateGmailIdentity] failed to get a refresh token for state %s. User should check application permissions.", state)
		return
	}
	// retrieve gmail profile from Google api
	httpClient := oauthConfig.Client(ctx, token)
	googleService, e := googleApi.New(httpClient)
	if e != nil {
		err = WrapCaliopenErr(e, NotFoundCaliopenErr, "[CreateGmailIdentity] failed to create google service")
		return
	}
	googleUser, e := googleService.Userinfo.Get().Do()
	if e != nil {
		err = WrapCaliopenErr(e, NotFoundCaliopenErr, "[CreateGmailIdentity] failed to retrieve user from google api")
		return
	}
	// build user identity
	// 1. check if this user_identity already exists
	foundIdentities, e := rest.store.LookupIdentityByIdentifier(googleUser.Email, EmailProtocol)
	if e != nil {
		err = WrapCaliopenErrf(e, DbCaliopenErr, "[CreateGmailIdentity] failed to lookup in store if identity already exists. Aborting")
		return
	}
	foundCount := len(foundIdentities)
	switch foundCount {
	case 0:
		userIdentity := new(UserIdentity)
		userID := UUID(uuid.FromStringOrNil(oauthCache.UserId))
		userIdentity.MarshallNew(userID)
		userIdentity.Protocol = EmailProtocol
		userIdentity.Type = RemoteIdentity
		userIdentity.DisplayName = googleUser.Name
		userIdentity.Identifier = googleUser.Email
		userIdentity.Credentials = &Credentials{
			users.CRED_ACCESS_TOKEN:  token.AccessToken,
			users.CRED_REFRESH_TOKEN: token.RefreshToken,
			users.CRED_TOKEN_EXPIRY:  token.Expiry.Format(time.RFC3339),
			users.CRED_TOKEN_TYPE:    token.TokenType,
			users.CRED_USERNAME:      googleUser.Email,
		}
		userIdentity.Infos = map[string]string{
			"inserver":  gmailProvider.Infos["imapserver"],
			"outserver": gmailProvider.Infos["smtpserver"],
			"provider":  "gmail",
			"authtype":  Oauth2,
		}

		// save identity
		e := rest.CreateUserIdentity(userIdentity)
		if e != nil {
			err = WrapCaliopenErr(e, FailDependencyCaliopenErr, "[CreateGmailIdentity] failed to create user identity")
			return
		}
		remoteId = userIdentity.Id.String()
	case 1:
		// this identity already exists,
		// WHAT TO DO ?? -> stop here or continue ?
		// checking if it belongs to this user and, if ok, just updating display name and tokens
		storedIdentity, e := rest.RetrieveUserIdentity(foundIdentities[0][0], foundIdentities[0][1], false)
		if e != nil || storedIdentity == nil {
			err = WrapCaliopenErrf(e, DbCaliopenErr, "[CreateGmailIdentity] failed to retrieve user identity found for google account %s", googleUser.Name)
			return
		}
		storedIdentity.DisplayName = googleUser.Name
		storedIdentity.Credentials = &Credentials{
			users.CRED_ACCESS_TOKEN:  token.AccessToken,
			users.CRED_REFRESH_TOKEN: token.RefreshToken,
			users.CRED_TOKEN_EXPIRY:  token.Expiry.Format(time.RFC3339),
			users.CRED_TOKEN_TYPE:    token.TokenType,
			users.CRED_USERNAME:      googleUser.Email,
		}

		modifiedFields := map[string]interface{}{
			"DisplayName": googleUser.Name,
			"Credentials": storedIdentity.Credentials,
		}
		if e := rest.store.UpdateUserIdentity(storedIdentity, modifiedFields); e != nil {
			err = WrapCaliopenErrf(e, FailDependencyCaliopenErr, "[CreateGmailIdentity] failed to update user identity in db")
			return
		}
		remoteId = storedIdentity.Id.String()
		return
	default:
		err = NewCaliopenErrf(FailDependencyCaliopenErr, "[CreateGmailIdentity] inconsistency in store : more than one identity found with email <%s>", googleUser.Email)
		return
	}
	return
}

func setTwitterAuthRequestUrl(provider *Provider, hostname string) (requestToken, requestSecret string, err CaliopenError) {

	provider.OauthCallbackUri = fmt.Sprintf(users.CALLBACK_BASE_URI, "twitter")

	//IMPORTANT TODO: make use of vault to store consumer key and secret
	conf := &oauth1.Config{
		ConsumerKey:    provider.Infos["consumer_key"],
		ConsumerSecret: provider.Infos["consumer_secret"],
		CallbackURL:    hostname + provider.OauthCallbackUri,
		Endpoint:       twitterOAuth1.AuthorizeEndpoint,
	}
	requestToken, requestSecret, e := conf.RequestToken()
	if e != nil {
		err = WrapCaliopenErrf(e, FailDependencyCaliopenErr, "[setTwitterAuthRequestUrl] failed with RequestToken()")
		return
	}
	authUrl, e := conf.AuthorizationURL(requestToken)
	if e != nil {
		err = WrapCaliopenErrf(e, FailDependencyCaliopenErr, "[setTwitterAuthRequestUrl] failed with AuthorizationURL()")
		return
	}
	provider.OauthRequestUrl = authUrl.String()
	return
}
