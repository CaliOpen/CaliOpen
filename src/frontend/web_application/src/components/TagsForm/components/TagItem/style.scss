@import '../../../../styles/common';
@import '../../../../styles/vendor/bootstrap_foundation-sites';

.m-tag {
  @include flex-grid-row($size: expanded);
  height: $co-component__height;
  margin-bottom: map-get($co-form__spacing, 'small');
  background: $co-color__fg__back--high;

  &__button {
    @include flex-grid-row($size: expanded);
    &:hover { background: $co-color__fg__back--higher;  }
  }

  &__text {
    @include flex-grid-size(shrink);
    line-height: $co-component__height;
  }

  &__icon {
    @include flex-grid-size(shrink);
    margin-left: auto;
    line-height: $co-component__height;
  }

  &__input { @include flex-grid-size; }
}
