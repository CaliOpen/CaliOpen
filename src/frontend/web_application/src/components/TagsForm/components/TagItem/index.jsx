import React, { Component, PropTypes } from 'react';
import Button from '../../../Button';
import Icon from '../../../Icon';
import { TextFieldGroup, FormGrid } from '../../../form';

import './style.scss';

function generateStateFromProps(props) {
  return { tag: { ...props.tag } };
}


class TagItem extends Component {
  static propTypes = {
    tag: PropTypes.shape({ type: PropTypes.string }).isRequired,
    onUpdate: PropTypes.func.isRequired,
  };

  constructor(props) {
    super(props);
    this.state = {
      tag: { name: '' },
      edit: false,
    };

    this.handleClickTag = this.handleClickTag.bind(this);
    this.handleUpdate = this.handleUpdate.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  componentWillMount() {
    this.setState(generateStateFromProps(this.props));
  }

  componentWillReceiveProps(newProps) {
    this.setState(generateStateFromProps(newProps));
  }

  handleClickTag() {
    this.setState(prevState => ({
      edit: !prevState.edit,
    }));
  }

  handleChange(ev) {
    const name = ev.target.value;

    this.setState(prevState => ({
      tag: { ...prevState.tag, name },
    }));
  }

  handleUpdate() {
    this.setState((prevState) => {
      this.props.onUpdate({ tag: this.state.tag, original: this.props.tag });

      return {
        tag: { ...prevState.tag, name },
      };
    });
  }

  renderForm() {
    const { tag } = this.props;

    return (
      <FormGrid className="m-tag">
        <TextFieldGroup
          name={tag.tag_id}
          className="m-tag__input"
          label={tag.name}
          placeholder={tag.name}
          value={this.state.tag.name}
          onChange={this.handleChange}
          showLabelforSr
          autoFocus
        />
        <Button inline onClick={this.handleUpdate}><Icon type="check" /></Button>
      </FormGrid>
    );
  }

  renderButton() {
    const { tag } = this.props;

    return (
      <FormGrid className="m-tag">
        <Button
          className="m-tag__button"
          onClick={this.handleClickTag}
          expanded
        >
          <span className="m-tag__text">{tag.name}</span>
          <Icon className="m-tag__icon" type="edit" />
        </Button>
      </FormGrid>
    );
  }

  renderSystem() {
    const { tag } = this.props;

    return (
      <div className="m-tag">
        <span className="m-tag__text">{tag.name}</span>
      </div>
    );
  }

  render() {
    if (this.state.edit) {
      return this.renderForm();
    }

    if (this.props.tag.type === 'system') {
      return this.renderSystem();
    }

    return this.renderButton();
  }
}

export default TagItem;
