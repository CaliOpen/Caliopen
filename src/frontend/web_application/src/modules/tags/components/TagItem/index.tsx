import React, { Component } from 'react';
import { withI18n, withI18nProps } from '@lingui/react';
import { Badge } from 'src/components';
import { getTagLabel } from '../../services/getTagLabel';

import './style.scss';
import { TagMixed } from '../../query';
import { TagPayload } from '../../types';

interface Props extends withI18nProps {
  tag: TagMixed;
  onDelete: (tag: TagMixed) => void | Promise<void>;
}
class TagItem extends Component<Props> {
  handleDeleteTag = async () => {
    const { onDelete, tag } = this.props;

    onDelete(tag);
  };

  render() {
    const { tag, i18n } = this.props;

    return (
      <Badge
        className="m-tag-item"
        onDelete={this.handleDeleteTag}
        ariaLabel={i18n._(/* i18n */ 'tags.action.remove', undefined, {
          message: 'Remove',
        })}
      >
        {tag.name ? getTagLabel(i18n, tag as TagPayload) : tag.label}
      </Badge>
    );
  }
}

export default withI18n()(TagItem);
