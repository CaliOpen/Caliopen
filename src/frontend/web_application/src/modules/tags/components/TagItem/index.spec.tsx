import * as React from 'react';
import { render } from '@testing-library/react';
import { AllProviders } from 'test/providers';
import TagItem from '.';
import { TagPayload } from '../../types';

describe('component TagItem', () => {
  it('render', () => {
    const tag: TagPayload = { name: 'foo', label: 'Foo', type: 'user' };
    const onDelete = () => {};

    const { container } = render(<TagItem tag={tag} onDelete={onDelete} />, {
      wrapper: AllProviders,
    });

    expect(container.firstChild).toMatchSnapshot();
  });
});
