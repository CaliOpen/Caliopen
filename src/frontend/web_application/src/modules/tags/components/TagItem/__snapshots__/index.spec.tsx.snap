// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component TagItem render 1`] = `
<div
  id="root"
>
  <span
    class="m-badge m-badge--normal-radius m-tag-item"
  >
    <span
      class="m-badge__text m-badge__text--has-button"
    >
      Foo
    </span>
    <button
      aria-label="Remove"
      class="m-badge__button m-button m-button--center m-button--inline m-button--text m-button--icon"
      type="button"
    >
      <i
        class="m-button__icon fa fa-remove"
      />
    </button>
  </span>
</div>
`;
